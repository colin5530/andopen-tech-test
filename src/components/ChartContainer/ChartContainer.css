.c-ChartContainer {
  margin: 10px 0;
}