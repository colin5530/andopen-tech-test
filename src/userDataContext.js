import React, { useContext, useReducer } from 'react';
import userData from './userData.json';

const UserDataContext = React.createContext();

const initialState = {
  totalUsers: userData.userData,
  visibleUsers: userData.userData,
  spendFilterValue: [0, 5000],
  regionFilterValue: [],
  genderFilterValue: [],
};

const applyFilters = ({totalUsers, spendFilterValue, regionFilterValue, genderFilterValue}) => {
  return totalUsers.filter((user) => {
    return user.spend >= spendFilterValue[0] && user.spend <= spendFilterValue[1];
  }).filter((user) => {
    if (regionFilterValue.length === 0) return true;
    return [...regionFilterValue].includes(user.region);
  }).filter((user) => {
    if (genderFilterValue === 'ALL') return true;
    return genderFilterValue === user.gender;
  });
}

const UserDataReducer = (state, action) => {
  switch (action.type) {
    case 'updateSpendFilter': {
      const filteredUsers = applyFilters(Object.assign({...state}, {spendFilterValue: action.payload}));

      return Object.assign({...state}, {visibleUsers: filteredUsers, spendFilterValue: action.payload})
    }
    case 'updateRegionFilter': {
      const filteredUsers = applyFilters(Object.assign({...state}, {regionFilterValue: action.payload}));

      return Object.assign({...state}, {visibleUsers: filteredUsers, regionFilterValue: action.payload})
    }
    case 'updateGenderFilter': {
      const filteredUsers = applyFilters(Object.assign({...state}, {genderFilterValue: action.payload}));

      return Object.assign({...state}, {visibleUsers: filteredUsers, genderFilterValue: action.payload})
    }
    case 'updateVisibleUsers': {
      return Object.assign({...state}, {visibleUsers: action.payload});
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`)
    }
  }
}

const UserDataProvider = ({children}) => {
  const [state, dispatch] = useReducer(UserDataReducer, initialState);
  const value = { state, dispatch };
  return <UserDataContext.Provider value={value}>{children}</UserDataContext.Provider>
}

const useUserData = () => {
  const context = useContext(UserDataContext);
  if (context === undefined) {
    throw new Error('useUserData must be used within a UserDataProvider');
  }
  return {visibleUsers: context.state.visibleUsers, setVisibleUsers: (payload) => context.dispatch({ type: 'updateVisibleUsers', payload})};
}

const useSpendFilter = () => {
  const context = useContext(UserDataContext);
  if (context === undefined) {
    throw new Error('useSpendFilter must be used within a UserDataProvider');
  }
  return {spendFilterValue: context.state.spendFilterValue, setSpendFilterValue: (payload) => context.dispatch({ type: 'updateSpendFilter', payload})};
}

const useRegionFilter = () => {
  const context = useContext(UserDataContext);
  if (context === undefined) {
    throw new Error('useRegionFilter must be used within a UserDataProvider');
  }
  return {regionFilterValue: context.state.regionFilterValue, setRegionFilterValue: (payload) => context.dispatch({ type: 'updateRegionFilter', payload})};
}

const useGenderFilter = () => {
  const context = useContext(UserDataContext);
  if (context === undefined) {
    throw new Error('useGenderFilter must be used within a UserDataProvider');
  }
  return {genderFilterValue: context.state.genderFilterValue, setGenderFilterValue: (payload) => context.dispatch({ type: 'updateGenderFilter', payload})};
}

export {
  UserDataProvider,
  useUserData,
  useSpendFilter,
  useRegionFilter,
  useGenderFilter,
}