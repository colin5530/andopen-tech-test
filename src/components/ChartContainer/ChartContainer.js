import React, { useEffect, useRef } from 'react';
import { CategoryScale, Chart, LineController, LineElement, PointElement, LinearScale, Title } from 'chart.js'
import { useUserData } from '../../userDataContext';

Chart.register(CategoryScale, LineController, LineElement, PointElement, LinearScale, Title);

const ChartContainer = () => {
  const chartRef = useRef(null);
  const {visibleUsers} = useUserData();
  
  useEffect(() => {
    const labels = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

    const monthlyData = labels.map((label, index) => {
      return visibleUsers.filter(user => user.birthday === (index + 1)).length;
    });

    const cumulativeData = monthlyData.map((sum => val => sum += (val * 5))(0));

    const chart = new Chart(chartRef.current.getContext('2d'), {
      type: 'line',
      data: {
          labels,
          datasets: [{
              label: '# of Users',
              data: monthlyData,
              backgroundColor: [
                  'rgba(255, 99, 132, 0.2)',
                  'rgba(54, 162, 235, 0.2)',
                  'rgba(255, 206, 86, 0.2)',
                  'rgba(75, 192, 192, 0.2)',
                  'rgba(153, 102, 255, 0.2)',
                  'rgba(255, 159, 64, 0.2)'
              ],
              borderColor: [
                  'rgba(255, 99, 132, 1)',
                  'rgba(54, 162, 235, 1)',
                  'rgba(255, 206, 86, 1)',
                  'rgba(75, 192, 192, 1)',
                  'rgba(153, 102, 255, 1)',
                  'rgba(255, 159, 64, 1)'
              ],
              borderWidth: 1
          }, {
            label: 'Cumulative',
            data: cumulativeData,
            backgroundColor: [
                'rgba(255, 99, 132, 0.2)',
                'rgba(54, 162, 235, 0.2)',
                'rgba(255, 206, 86, 0.2)',
                'rgba(75, 192, 192, 0.2)',
                'rgba(153, 102, 255, 0.2)',
                'rgba(255, 159, 64, 0.2)'
            ],
            borderColor: [
                'rgba(255, 99, 132, 1)',
                'rgba(54, 162, 235, 1)',
                'rgba(255, 206, 86, 1)',
                'rgba(75, 192, 192, 1)',
                'rgba(153, 102, 255, 1)',
                'rgba(255, 159, 64, 1)'
            ],
            borderWidth: 1
        }]
      },
    });

    return () => chart.destroy();
  }, [visibleUsers]);

  return (
    <div>
      Chart
      <canvas ref={chartRef} id="myChart" width="400" height="200"></canvas>
    </div>
  )
};

export default ChartContainer;