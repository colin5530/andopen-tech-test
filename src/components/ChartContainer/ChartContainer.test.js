import renderer from 'react-test-renderer';
import ChartContainer from './';
import { UserDataProvider } from '../../userDataContext';

describe('ChartContainer', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <UserDataProvider>
        <ChartContainer />
      </UserDataProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
