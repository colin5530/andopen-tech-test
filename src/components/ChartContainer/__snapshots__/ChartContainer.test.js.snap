// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartContainer renders correctly 1`] = `
<div
  className="c-ChartContainer"
>
  <canvas
    data-testid="canvas"
    height={150}
    onClick={[Function]}
    role="img"
    width={300}
  />
</div>
`;
